pyadjoint/checkpointing.py

```diff
@@ -0,0 +1,305 @@
+from enum import Enum
+from functools import singledispatchmethod
+from checkpoint_schedules import (Copy, Move, EndForward, EndReverse, Forward, Reverse, StorageType)
+from checkpoint_schedules import Revolve
+
+
+class CheckpointError(RuntimeError):
+    pass
+
+
+class Mode(Enum):
+    """The mode of the checkpoint manager."""
+    RECORD = 1
+    FINISHED_RECORDING = 2
+    EVALUATED = 3
+    EXHAUSTED = 4
+    RECOMPUTE = 5
+    EVALUATE_ADJOINT = 6
+
+
+class AdjointSchedule(list):
+    """A list of checkpoint actions with a step count.
+
+    Attributes
+    ----------
+    steps : int
+        The number of forward steps in the initial forward calculation.
+    """
+    def __init__(self, cp_action=(), steps=0):
+        # Steps is only intended to provide a rough indicator for progress
+        # bars.
+        super().__init__(cp_action)
+        self.steps = steps
+
+
+def process_schedule(schedule):
+    """Process a checkpoint schedule into forward and adjoint schedules.
+
+    Parameters
+    ----------
+    schedule : checkpoint_schedules.schedule
+        A schedule provided by the checkpoint_schedules package.
+
+    Returns
+    -------
+    list, list
+        Forward and adjoint list of schedules.
+    """
+    schedule = list(schedule)
+    index = 0
+    forward_steps = 0
+
+    while not isinstance(schedule[index], EndForward):
+        if isinstance(schedule[index], Forward):
+            forward_steps += len(schedule[index])
+        index += 1
+    end_forward = index + 1
+    reverse_steps = 0
+
+    while not isinstance(schedule[index], EndReverse):
+        if isinstance(schedule[index], Reverse) or isinstance(schedule[index], Forward):  # noqa: E501
+            reverse_steps += len(schedule[index])
+        index += 1
+    forward = AdjointSchedule(schedule[:end_forward], forward_steps)
+    reverse = AdjointSchedule(schedule[end_forward:], reverse_steps)
+
+    return forward, reverse
+
+
+class CheckpointManager:
+    """Manage the executions of the forward and adjoint solvers.
+
+    Attributes
+    ----------
+    schedule : checkpoint_schedules.schedule
+        A schedule provided by the checkpoint_schedules package.
```

Comment:
Should mention that only `Revolve` is currently supported. Otherwise users only discover this when they run their code.